Feature implement WeatherService

MarsLviv opened the pull request.

MarsLviv committed:
add WeatherService class

MarsLviv committed:
add WeatherService connection into MainActivity

MarsLviv committed:
add WeatherRequest class

MarsLviv committed:
onError method implemtntation

MarsLviv committed:
remove obsolete code from MainActivity class

MarsLviv committed:
some minor changes

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -48,6 +31,9 @@
     private LocationClient locationClient;
     private Location currentLocation;
 
+    WeatherService weatherService;
```
```suggestion
    private WeatherService weatherService;
```

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -48,6 +31,9 @@
     private Location currentLocation;
 
+    WeatherService weatherService;
+    boolean weatherServiceBound = false;
```
```suggestion
```

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -70,59 +56,63 @@ public void onRetrieveLocation(@NonNull Location location) {
+
+        @Override
+        public void onServiceDisconnected(ComponentName arg0) {
+            weatherServiceBound = false;
```
```suggestion
            weatherService = null;
```

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -70,59 +56,63 @@ public void onRetrieveLocation(@NonNull Location location) {
+        super.onStop();
+
+        unbindService(connection);
+        weatherServiceBound = false;
```
```suggestion
```

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -70,59 +56,63 @@ public void onRetrieveLocation(@NonNull Location location) {
+        public void onServiceConnected(ComponentName className, IBinder service) {
+            WeatherService.WeatherBinder binder = (WeatherService.WeatherBinder) service;
+            weatherService = binder.getService();
+            weatherServiceBound = true;
```
```suggestion
```

gutachkee commented on app/src/main/java/com/example/weather/WeatherService.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public interface WeatherServiceCallback {
+        void onWeatherInfoObtained(Optional<WeatherInfo> weatherInfo);
```
```suggestion
        void onWeatherInfoObtained(@Nonnull WeatherInfo weatherInfo);
```

gutachkee commented on app/src/main/java/com/example/weather/WeatherRequest.java:
```diff
@@ -0,0 +1,108 @@
+
+    private final Resources resources;
+
+    public WeatherRequest (@NonNull Resources resources) {
```
```suggestion
    public WeatherRequest(@NonNull Resources resources) {
```

MarsLviv committed:
Refactoring:
class MainActivity: -move variable connection to section with variables;
-remove bool variable weatherServiveBound;
interface WeatherServiceCallback: -change onWeatherInfoObtained signature

MarsLviv committed:
Refactor:

Information that need for https request (cityName or Location, API end point, API key) covered into class WeatherRequestData.
Object of this class passes into WeatherService and WeatherRequest constractors.

MarsLviv committed:
Refactoring:

- class WearherService: -rename inner class to WetherServiceRunnable;
- add variable visibility modifiers;
- merged two try/catch block into one.

MarsLviv committed:
add UML dir

MarsLviv committed:
Added sequenseDiagram.drawio

MarsLviv committed:
Refactoring: class MainActivity: -rename metod

MarsLviv committed:
Merge branch 'feature-implement_WeatherService' of https://github.com/android-on-MarsLviv/Weather into feature-implement_WeatherService

MarsLviv committed:
Update SequenceDiagram.drawio

MarsLviv committed:
Refactoring:

-renwming

MarsLviv committed:
Update SequenseDiagram.drawio

MarsLviv committed:
Refactoring:
-builder make uri depending on city name or location;
-WeatherInfoProvider provide just one public method for clients code

MarsLviv committed:
Merge branch 'feature-implement_WeatherService' of https://github.com/android-on-MarsLviv/Weather into feature-implement_WeatherService

MarsLviv committed:
Refactoring: add some minor changes

MarsLviv committed:
class WeatherInfo: add NonNull identifier

MarsLviv committed:
minor changes

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -70,61 +68,47 @@ public void onRetrieveLocation(@NonNull Location location) {
-        // todo: use Executors.newSingleThreadExecutor() executor to run network task on a separate thread.
-        // https://trello.com/c/O0wDKeQP
-        new Thread(new Runnable() {
+        WeatherRequest weatherRequest = new WeatherRequest.Builder(getText(R.string.weather_api_key).toString(), getText(R.string.weather_api_entry_point).toString())
+                .setCity(cityName)
+                .build();
```
:+1:

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -139,7 +123,7 @@ public void onRequestPermissionsResult(int requestCode, @NonNull String[] permis
     }
 
-    private void updateTemperatureView(String massage) {
+    private void updateWeatherView(@NonNull String massage) {
```
```suggestion
    private void updateWeatherView(@NonNull String message) {
```

gutachkee commented on app/src/main/java/com/example/weather/MainActivity.java:
```diff
@@ -148,80 +132,8 @@ public void run() {
-
-    private void notificationOnError(String notificationToUser) {
-        updateTemperatureView(getText(R.string.default_weather_message).toString());
+    private void notificationOnError(@NonNull String notificationToUser) {
```
```suggestion
    private void notificationOnError(@NonNull String errorMessage) {
```

gutachkee commented on app/src/main/java/com/example/weather/WeatherInfo.java:
```diff
@@ -1,12 +1,14 @@
     private final String windSpeed;
 
-    WeatherInfo(String temperature, String humidity, String visibility, String windSpeed) {
+    WeatherInfo(@NonNull String temperature, @NonNull String humidity, @NonNull String visibility, @NonNull String windSpeed) {
```
Add @Nonnull annotation to getters too

gutachkee commented on app/src/main/java/com/example/weather/WeatherInfoProvider.java:
```diff
@@ -0,0 +1,112 @@
+            WeatherInfo weather = weatherInfo.get();
+            callback.onRequestSucceed(weather);
+        } catch (IOException e) {
+            e.printStackTrace();
```
```suggestion
            e.printStackTrace();
            callback.onRequestFailed();
```

gutachkee commented on app/src/main/java/com/example/weather/WeatherRequest.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final String cityName;
+    private final Location location;
+
```
```suggestion
```

gutachkee commented on app/src/main/java/com/example/weather/WeatherService.java:
```diff
@@ -0,0 +1,80 @@
+                @Override
+                public void onRequestFailed() {
+                    Log.i(TAG, "RequestFailed");
+                    callback.onError(new Error(""));
```
Re-define onRequestFailed() method to provide error code or message. So you can pass it as result for the user.

gutachkee commented on app/src/main/java/com/example/weather/WeatherService.java:
```diff
@@ -0,0 +1,80 @@
+
+    public interface WeatherServiceCallback {
+        void onWeatherInfoObtained(@NonNull WeatherInfo weatherInfo);
+        void onError(@NonNull Error errorCode);
```
```suggestion
        void onError(@NonNull Error error);
```

gutachkee reviewed (commented):
Great work done!
More cosmetic comments that architecture-related.

MarsLviv committed:
disable button during request

MarsLviv committed:
Refactoring:
-variable renaming;
-class WeatherInfoProvider: add callback.onRequestFailed in catch part

MarsLviv committed:
class WeatherService: change onError interface

MarsLviv committed:
Refactoring:

-WeatherService: relocate callback calling into MainActivity with Handler;
-MainActivity: remove unused methods updateWeatherView, notificationOnError.

gutachkee commented on app/src/main/java/com/example/weather/WeatherService.java:
```diff
@@ -0,0 +1,95 @@
+                    handler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            callback.onError();;
```
```suggestion
                            callback.onError();
```
Do self-review to eliminate cosmetic issues.

gutachkee commented on app/src/main/java/com/example/weather/WeatherService.java:
```diff
@@ -0,0 +1,95 @@
+
+                @Override
+                public void onRequestFailed() {
+                    handler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            callback.onError();;
+                        }
+                    });
```
Learn how to use lambdas and use them. Code looks much cleaner. Create a ticket (in trello) if you need 
```suggestion
                    handler.post(() -> callback.onError());
```

gutachkee reviewed (approved):
few cosmetic change request. After fixing them feel free to merge this PR. 
Thanks
